Source/Core/VideoCommon/LightingShaderGen.h

```diff
@@ -62,7 +62,8 @@ static void GenerateLightShader(T& object, LightingUidData& uid_data, int index,
 		case LIGHTATTN_NONE:
 		case LIGHTATTN_DIR:
 			object.Write("ldir = normalize(" LIGHT_POS".xyz - pos.xyz);\n", LIGHT_POS_PARAMS(index));
-			object.Write("attn = 1.0f;\n");
+			object.Write("attn = 1.0;\n");
+			object.Write("if (length(ldir) == 0.0)\n\t ldir = float3(1.0, 1.0, 1.0);\n");
```

Comment:
No, that's not true, we do get predictable results.  The problem is xfmem is giving us Light Position data with 0 and near infinity values e+019.  After normalizing the Light Direction we get an overflow in Length2().  The normalized ldir ends up as Vec3(0,0,0) or ldir.Length() = 0.0f giving us a reliable way to detect an overflowed ldir when normalized.  This code works because of this detection. The other way we could do it is to do this instead first:
if (light->pos.Length() == std::numeric_limits<float>::infinity())

Either way, the results are the same.